llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -5797,7 +5797,7 @@ SDValue DAGCombiner::hoistLogicOpWithSameOpcodeHands(SDNode *N) {
   }
 
   // logic_op (truncate x), (truncate y) --> truncate (logic_op x, y)
-  if (HandOpcode == ISD::TRUNCATE) {
+  if (HandOpcode == ISD::TRUNCATE && !TLI.shouldReduceRegisterPressure()) {
```

Comment:
I've added a comment. 

I added support for `TLI.isTruncateFree(EVT, EVT)`. 

I'm holding off on implementing `TLI.ZExtFree()` because it isn't necessary for this optimization.

Reply:
TODO: File an issue for this.

Reply:
#114339